### DISCUSSION TITLE
Redirecting user to /api/auth/signin/[PROVIDER] results in an error?

### DISCUSSION THREAD

**Discussion by liamlows** [nextauthjs/next-auth#11220] [Help]
We are currently facing an issue where we cannot redirect the user directly to the provider login page. Currently we have Keycloak set up as a provider and if the user clicks on a button that triggers `signIn('keycloak', { callbackUrl: '/dashboard' })` everything works as intended.

To add full site protection on most pages, we are using middleware with the `auth()` function. In that function if `auth.req` is null or undefined we have a redirect setup as follows so that the user will be redirected to the login page and then sent back to the page they intended to visit:

```
// middleware.ts
return NextResponse.redirect(
  new URL(
    `/api/auth/signin/keycloak?callbackUrl=${encodeURIComponent(`${pathname}${search}`)}`,
    origin,
  ),
);
```
 
Unfortunately, this does not work as when the middleware redirects the user to the link above, it errors with `UnknownAction: Unsupported action.` We found two workarounds:

1. Send the user redirect to `api/auth/signin` and they then can click the "login with keycloak" button rendered by next-auth. (This is unideal as it introduces an unnecessary click to get the user to the login form)

2. Creating an api route that calls the `signin()` function with the callback url from the original request.

```
// api/signin/route.ts
import { signIn } from '../../../auth';

export async function GET(req: Request) {
  const searchParams = new URL(req.url).searchParams;
  return signIn('keycloak', { redirectTo: searchParams.get('callbackUrl') ?? '' });
}
```
```
// middleware.ts
return NextResponse.redirect(
  new URL(
    `/api/signin?callbackUrl=${encodeURIComponent(`${pathname}${search}`)}`,
    origin,
  ),
);
```

We chose option 2 and it works however for some reason feels a bit wrong. With other OIDC providers like auth0 we can just send the user directly to the signin page without having to the interception you see above.

Is this the best approach or is there a better way to send the user straight to provider login?

**Comment by carlosrodovalho** [comment]
i have the same issue. do you managed to get the solution?

**Comment by showduhtung** [comment]
I do think next-auth is a great solution and all, but with everything in a blackbox, it's incredibly difficult to debug something simple like this.

**Comment by wingsofovnia** [comment]
Here is a more generic solution

`cat > app/api/auth/signin/[provider]/route.ts <<EOF`
```typescript
import { type NextRequest } from 'next/server'

import { signIn } from '@/auth.ts'

/**
 * Handles GET requests to sign in with a specific authentication provider.
 *
 * This route handler provides a route `GET /api/auth/signin/{provider}`
 * to sign in with a specific authentication provider, which is otherwise
 * not possible with a GET request in Auth.js out of the box, useful for 
 * redirects in {@link authOptions.callbacks.authorized} callbacks.
 */
export async function GET(
  req: NextRequest,
  { params: { provider } }: { params: { provider?: string } },
) {
  const callback = req.nextUrl.searchParams.get('callback')
  return signIn(provider, callback ? { redirectTo: callback } : undefined)
}
```
`EOF`

This will enable  `GET /api/auth/signin/[provider]`, such as `/api/auth/signin/keycloak` or `/api/auth/signin/apple`.

You might also want to avoid hardcoding the provider in `authorized` callback and instead redirect straight to the provider if there is only one provider registered in Auth.js, otherwise go to default signin page:

```typescript
import NextAuth, { NextAuthConfig } from 'next-auth'
import { NextResponse } from 'next/server'
import Keycloak from '@auth/core/providers/keycloak'

export const authOptions: NextAuthConfig = {
  providers: [
    Keycloak({
      /* Params Here */
    }),
  ],
  callbacks: {
    authorized: async({ auth, request: { nextUrl } }) => {
      const isAuthorized = auth?.user && !auth.accessToken.isExpired // Replace with your auth check
      if (isAuthorized) return true

      const providerId = authOptions.providers.length == 1
          ? authOptions.providers
            .filter(p => 'id' in p)
            .map(p => p.id)[0] ?? ''
          : ''

      const redirectUrl = new URL(`/api/auth/signin/${providerId}`, nextUrl.origin)
      redirectUrl.searchParams.append('callbackUrl', nextUrl.href)

      return NextResponse.redirect(redirectUrl)
    },
  },
}

export const { handlers, signIn, signOut, auth } = NextAuth(authOptions)

```
